apps/heft/src/plugins/TypeScriptPlugin/TypeScriptBuilder.ts

```diff
@@ -494,120 +512,133 @@ export class TypeScriptBuilder extends SubprocessRunnerBase<ITypeScriptBuilderCo
-          const filename: string = sourceFile.fileName;
-          if (typeScriptFilenames.has(filename)) {
-            const relativeFilenameWithoutExtension: string = ts.removeFileExtension(
-              ts.getExternalModuleNameFromPath(resolverHost, filename)
-            );
+        this._printDiagnosticMessage(ts, diagnostic, diagnosticCategory);
+      }
+    }
 
-            for (const { cacheOutFolderPath, outFolderPath, jsExtensionOverride = '.js', isPrimary } of this
-              ._moduleKindsToEmit) {
-              // Only primary module kinds emit declarations
-              if (isPrimary) {
-                if (tsconfig.options.declarationMap) {
-                  const dtsMapFilename: string = `${relativeFilenameWithoutExtension}.d.ts.map`;
-                  queueLinkOrCopy({
-                    linkTargetPath: path.join(cacheOutFolderPath, dtsMapFilename),
-                    newLinkPath: path.join(outFolderPath, dtsMapFilename)
-                  });
-                }
-
-                if (tsconfig.options.declaration) {
-                  const dtsFilename: string = `${relativeFilenameWithoutExtension}.d.ts`;
-                  queueLinkOrCopy({
-                    linkTargetPath: path.join(cacheOutFolderPath, dtsFilename),
-                    newLinkPath: path.join(outFolderPath, dtsFilename)
-                  });
-                }
-              }
-
-              if (tsconfig.options.sourceMap && !sourceFile.isDeclarationFile) {
-                const jsMapFilename: string = `${relativeFilenameWithoutExtension}${jsExtensionOverride}.map`;
-                queueLinkOrCopy({
-                  linkTargetPath: path.join(cacheOutFolderPath, jsMapFilename),
-                  newLinkPath: path.join(outFolderPath, jsMapFilename)
-                });
-              }
-
-              // Write the .js file last in case something is watching its timestamp
-              if (!sourceFile.isDeclarationFile) {
-                const jsFilename: string = `${relativeFilenameWithoutExtension}${jsExtensionOverride}`;
-                queueLinkOrCopy({
-                  linkTargetPath: path.join(cacheOutFolderPath, jsFilename),
-                  newLinkPath: path.join(outFolderPath, jsFilename)
-                });
-              }
-            }
-          }
-        }
+    if (eslint) {
+      eslint.reportFailures();
+    }
+
+    if (tslint) {
+      tslint.reportFailures();
+    }
+
+    if (typeScriptErrorCount > 0) {
+      throw new Error(`Encountered TypeScript error${typeScriptErrorCount > 1 ? 's' : ''}`);
+    }
+  }
+
+  public async _runSolutionBuild(
+    ts: ExtendedTypeScript,
+    measureTsPerformance: PerformanceMeasurer
+  ): Promise<void> {
+    // Ensure the cache folder exists
+    this._cachedFileSystem.ensureFolder(this._configuration.buildCacheFolder);
+
+    this._typescriptTerminal.writeVerboseLine(`Using solution mode`);
+
+    //#region CONFIGURE
```

Comment:
There are a handful of discrepancies between this and the `_runBuild` "CONFIGURE" region. Is that because some of the things the other one does don't apply to the solution builder.

It seems logical that this statement:

```TypeScript
    if (this._configuration.watchMode) {
      await this._runWatch(ts, measureTsPerformance);
    } else if (this._useSolutionBuilder) {
      await this._runSolutionBuild(ts, measureTsPerformance);
    } else {
      await this._runBuild(ts, measureTsPerformance, measureTsPerformanceAsync);
    }
```

Would actually look like this:

```TypeScript
    if (this._configuration.watchMode) {
      await this._runWatch(ts, measureTsPerformance);
    } else {
      await this._runBuild(ts, measureTsPerformance, measureTsPerformanceAsync);
    }
```

and then have `runWatch` and `_runBuild` do similar configuration regardless of whether the solution builder is being used, and then use different behavior when running the actual build.

Reply:
Unfortunately some of the hooks in the pipeline (namely async file writes, and logging performance as we go) make the `_runBuildAsync` flow largely incompatible with the `_runSolutionBuildAsync` flow. I've at least refactored some of the common logic into helper functions to reduce duplication.